File: code/Project_4/BT13/dfs_general_graph.py
import queue
from collections import deque

def read_general_graph(filename):
    with open(filename, 'r') as f:
        lines = f.readlines()

    n, m = map(int, lines[0].split())
    adj = [[] for _ in range(n)]
    for i in range(1, m + 1):
        u, v = map(int, lines[i].split())
        adj[u].append(v)
        if u != v:
            adj[v].append(u)  # vô hướng

    for neighbors in adj:
        neighbors.sort()

    return n, adj

def dfs_general_graph(n, adj):
    visited = [False] * n
    parent = [-1] * n
    dfs_forest = []

    def dfs(u):
        visited[u] = True
        dfs_forest.append(u)
        for v in adj[u]:
            if not visited[v]:
                parent[v] = u
                dfs(v)

    for u in range(n):
        if not visited[u]:
            dfs(u)

    print("DFS forest order:", dfs_forest)
    print("Parent array:", parent)
    return dfs_forest, parent


if __name__ == "__main__":
    n, adj = read_general_graph("general_graph.inp")
    dfs_general_graph(n, adj)

File: code/Project_4/BT13/dfs_general_graph.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <string>
#include <algorithm>

using namespace std;

void read_general_graph(const string& filename, int& n, vector<vector<int>>& adj) {
    ifstream fin(filename);
    if (!fin) {
        cerr << "Cannot open file: " << filename << endl;
        exit(1);
    }
    int m;
    fin >> n >> m;
    adj.assign(n, vector<int>());
    for (int i = 0; i < m; ++i) {
        int u, v;
        fin >> u >> v;
        adj[u].push_back(v);
        if (u != v) adj[v].push_back(u); // vô hướng
    }
    fin.close();

    for (int i = 0; i < n; ++i) {
        sort(adj[i].begin(), adj[i].end());
    }
}

void dfs_from(int u, const vector<vector<int>>& adj, vector<bool>& visited, vector<int>& parent, vector<int>& dfs_order) {
    visited[u] = true;
    dfs_order.push_back(u);
    for (int v : adj[u]) {
        if (!visited[v]) {
            parent[v] = u;
            dfs_from(v, adj, visited, parent, dfs_order);
        }
    }
}

int main() {
    int n;
    vector<vector<int>> adj;
    read_general_graph("general_graph.inp", n, adj);

    vector<bool> visited(n, false);
    vector<int> parent(n, -1);
    vector<int> dfs_order;

    for (int i = 0; i < n; ++i) {
        if (!visited[i]) {
            dfs_from(i, adj, visited, parent, dfs_order);
        }
    }

    cout << "DFS forest order (general graph): ";
    for (int v : dfs_order) cout << v << " ";
    cout << "\nParent array: ";
    for (int p : parent) cout << p << " ";
    cout << "\n";

    return 0;
}
